### PR TITLE
Add an input property to the execute resource for passing input on STDIN

### PR DESCRIPTION
**Pull request by phiggins**
Signed-off-by: Pete Higgins <pete@peterhiggins.org>

Given an `execute` resource like this:

```
Petes-MBP:chef pete$ cat test_execute_stdin.rb
execute "wc -l > test_execute_stdin_output" do
  input <<-INPUT
    1
    2
    3
    4
    5
  INPUT
end
```

... the `input` property is passed to the command on STDIN:

```
Petes-MBP:chef pete$ chef-apply test_execute_stdin.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * execute[wc -l > test_execute_stdin_output] action run
    - execute wc -l > test_execute_stdin_output
Petes-MBP:chef pete$ cat test_execute_stdin_output
       5
```

`mixlib-shellout` has [a pretty dire warning](https://github.com/chef/mixlib-shellout/blob/master/lib/mixlib/shellout.rb#L143-L148) about doing this, but I tested on my local macOS environment, an ubuntu 18.04 VM, and a Windows 2019 VM and didn't notice any problems.

I made a test like this:

```
require "minitest/autorun"
require "digest"

describe "large script" do
  let(:data) { DATA.read }

=begin
Petes-MBP:chef pete$ seq 1 200000 | md5sum
0e10426a1d5bddffcef02f1345787128  -
=end
  let(:expected_md5) { "0e10426a1d5bddffcef02f1345787128" }

  it "reads embedded data correctly" do
    assert_equal data.split("\n").length, 200_000
    assert_equal Digest::MD5.hexdigest(data), expected_md5
  end
end

# seq 1 200000 >> test_large_stdin.rb
__END__
```

and then appended the integers from 1 to 200k to the DATA segment. This makes the script a little over one megabyte:
```
Petes-MBP:chef pete$ ls -lh test_large_stdin.rb
-rw-rw-rw-  1 pete  staff   1.2M May 22 16:01 test_large_stdin.rb
```

This was able to run with `Mixlib::ShellOut` on all 3 mentioned OSes with no noticeable differences.

```
Petes-MBP:chef pete$ sw_vers
ProductName:    Mac OS X
ProductVersion: 10.15.4
BuildVersion:   19E287
Petes-MBP:chef pete$ ruby -rmixlib/shellout -e 's = Mixlib::ShellOut.new("ruby", input: File.read("test_large_stdin.rb")) ; s.run_command ; s.error! ; puts s.stdout'
Run options: --seed 48632

# Running:

.

Finished in 0.057283s, 17.4572 runs/s, 34.9144 assertions/s.

1 runs, 2 assertions, 0 failures, 0 errors, 0 skips
```

```
vagrant@ubuntu-bionic:/vagrant/chef$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.4 LTS
Release:        18.04
Codename:       bionic
vagrant@ubuntu-bionic:/vagrant/chef$ ruby -rmixlib/shellout -e 's = Mixlib::ShellOut.new("ruby", input: File.read("test_large_stdin.rb")) ; s.run_command ; s.error! ; puts s.stdout'
Run options: --seed 49161

# Running:

.

Finished in 0.039495s, 25.3196 runs/s, 50.6391 assertions/s.

1 runs, 2 assertions, 0 failures, 0 errors, 0 skips
```

```
PS C:\vagrant\chef> systeminfo | grep "^OS Name\|^OS Version"
OS Name:                   Microsoft Windows Server 2019 Datacenter Evaluation
OS Version:                10.0.17763 N/A Build 17763
PS C:\vagrant\chef> ruby -rmixlib/shellout -e "s = Mixlib::ShellOut.new('ruby', input: File.read('test_large_stdin.rb')) ; s.run_command ; s.error! ; puts s.stdout"
Run options: --seed 4336

# Running:

.

Finished in 0.055300s, 18.0831 runs/s, 36.1662 assertions/s.

1 runs, 2 assertions, 0 failures, 0 errors, 0 skips
```

### CODE REVIEW

File: lib/chef/resource/execute.rb [hunk 1]
```diff
@@ -103,6 +103,10 @@ def initialize(name, run_context = nil)
         description: "Determines whether the script will run with elevated permissions to circumvent User Access Control (UAC) interactively blocking the process.\nThis will cause the process to be run under a batch login instead of an interactive login. The user running #{Chef::Dist::CLIENT} needs the 'Replace a process level token' and 'Adjust Memory Quotas for a process' permissions. The user that is running the command needs the 'Log on as a batch job' permission.\nBecause this requires a login, the user and password properties are required.",
         introduced: "13.3"
 
+      property :input, [String, nil],
+        introduced: "16.1",
```
Comment (tas50) [inline lib/chef/resource/execute.rb:107]: 16.2

File: lib/chef/resource/execute.rb [hunk 2]
```diff
@@ -103,6 +103,10 @@ def initialize(name, run_context = nil)
         description: "Determines whether the script will run with elevated permissions to circumvent User Access Control (UAC) interactively blocking the process.\nThis will cause the process to be run under a batch login instead of an interactive login. The user running #{Chef::Dist::CLIENT} needs the 'Replace a process level token' and 'Adjust Memory Quotas for a process' permissions. The user that is running the command needs the 'Log on as a batch job' permission.\nBecause this requires a login, the user and password properties are required.",
         introduced: "13.3"
 
+      property :input, [String, nil],
```
Comment (tas50) [inline lib/chef/resource/execute.rb:106]: Just string. Nil is the default for strings

**Comment by tas50** [comment]
Can you add an example to the examples block that shows how someone might use this so we can make sure it shows up on the docs site.

**Comment by jaymzh** [comment]
Agreed on docs. Also I'm a huge :+1: - it's a PR I've been meaning to write forever.

**Comment by lamont-granquist** [comment]
uh shit, we probably need to fix mixlib-shellout if that comment is right.  shouldn't block on this since nothing is using it yet, but that's terrible if that is correct.

File: lib/chef/resource/execute.rb [hunk 3]
```diff
@@ -27,10 +27,542 @@ class Execute < Chef::Resource
...
+          user "username@domain-name"
+          password "password"
+        end
+        ```
+
+        **Run a command with an external input file**:
+
+        execute 'md5sum' do
+          input File.read(__FILE__)
+        end
```
Comment (phiggins) [inline lib/chef/resource/execute.rb:564]: I couldn't think of an actual useful example of this.  ¯\\\_(ツ)\_/¯

**Pull request merged**